### PR TITLE
Feature 12030 add create edit form for environment on mobile

### LINKED ISSUE #12030: Add a create/edit form for environments in the mobile app [5]
### Feature Description

After an Environment entity has been created in https://github.com/SORMAS-Foundation/SORMAS-Project/issues/11796 as part of epic https://github.com/SORMAS-Foundation/SORMAS-Project/issues/11533, create and edit forms for this entity are now required for the mobile app.

### Added Value/Benefit

Users will be able to create and edit environment entities for further usage in the system.

### Acceptance Criteria

- [ ] A creation form for environments has been added with the following fields:
  - Report date (required)
  - External ID
  - Environment name (required)
  - Environment media (required)
  - Location form (region, district, gps lat and gps lon required; region and district read-only and pre-filled with the jurisdiction of the user)
  - Responsible user (contains a list of all users with the `ENVIRONMENT_EDIT` user right that are assigned to the district selected in the location form)
- [ ] An edit form for environments has been added with all fields from the creation form and the following additional fields:
  - UUID (read only)
  - Reporting user (read only)
  - Investigation status (required)
  - Water type (if environment media == water)
  - Other water type (if water type == other)
  - Infrastructure details (if environment media == water)
  - Other infrastructure details (if infrastructure details == other)
  - A "Water use" heading (if environment media == water)
  - Checkboxes for all `WaterUse` enum values (if environment media == water); see epidemiologicalEvidenceDetails of events in the web app for reference - this hasn't been implemented in the mobile app so far
  - Other water use (if the checkbox for "Other" water use has been checked)
- [ ] A button to create a new environment has been added to the environment list created in #12025 that brings up the creation form
- [ ] Clicking on an environment in the environment list created in #12025 takes the user to a dedicated page for the environment, displaying the edit form
- [ ] The location form includes the functionality to pick the current GPS coordinates of the user

### Implementation Details

_No response_

### Mockups

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by sergiupacurariu**
Fixes #12030

### CODE REVIEW

File: sormas-api/src/main/resources/captions.properties [hunk 1]
```diff
@@ -1123,6 +1123,7 @@ Environment.otherInfrastructureDetails=Other infrastructure details
 Environment.waterUse=Water use
 Environment.otherWaterUse=Other water use
 Environment.location=Location
+Environment.facilityDetails=
```
Comment (leventegal-she) [inline sormas-api/src/main/resources/captions.properties:1126]: Missing value

File: sormas-app/app/src/main/java/de/symeda/sormas/app/backend/environment/Environment.java [hunk 2]
```diff
@@ -195,22 +196,25 @@ public void setWaterUseJson(String waterUseJson) {
...
-            Gson gson = new Gson();
-            Type type = new TypeToken<Map<WaterUse, Boolean>>() {
-            }.getType();
-            wateruse = gson.fromJson(waterUseJson, type);
-            if (wateruse == null) {
-                wateruse = new HashMap<>();
-            }
-        }
+		if (wateruse == null) {
+			Gson gson = new GsonBuilder().enableComplexMapKeySerialization().registerTypeAdapter(WaterUse.class, new WaterUseSerializer()).create();
```
Comment (leventegal-she) [inline sormas-app/app/src/main/java/de/symeda/sormas/app/backend/environment/Environment.java:200]: please introduce a method for creating the Gson instance

File: sormas-app/app/src/main/java/de/symeda/sormas/app/backend/environment/EnvironmentDao.java [hunk 3]
```diff
@@ -84,4 +97,19 @@ public void deleteEnvironmentAndAllDependingEntities(String environmentUuid) thr
 		// Delete case
 		deleteCascade(environment);
 	}
+
+	@Override
+	public Environment saveAndSnapshot(final Environment environment) throws DaoException {
+		// If a new environment is created, use the last available location to update its report latitude and longitude
```
Comment (leventegal-she) [inline sormas-app/app/src/main/java/de/symeda/sormas/app/backend/environment/EnvironmentDao.java:103]: I'm not sure this is needed, is it requested by the ticket?

File: sormas-app/app/src/main/java/de/symeda/sormas/app/component/controls/ControlCheckBoxGroupFieldForMap.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+				}
+
+				checkBox.setChecked(element.getValue());
+			}
+		}
+	}
+
+	@Override
+	public void setEnabled(boolean enabled) {
+		// TODO: Implement; Probably replace elements with tags
```
Comment (leventegal-she) [inline sormas-app/app/src/main/java/de/symeda/sormas/app/component/controls/ControlCheckBoxGroupFieldForMap.java:162]: this logic from `setReadonly` should go here and there will be no need for that method

File: sormas-app/app/src/main/java/de/symeda/sormas/app/component/controls/ControlCheckBoxGroupFieldForMap.java [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+		"enumClass" })
+	public static <E extends Enum<?>> void setValue(ControlCheckBoxGroupFieldForMap<E> view, Map<E, Boolean> value, Class<E> enumClass) {
+		if (enumClass != null) {
+			view.setEnumClass(enumClass);
+		}
+		view.setFieldValue(value);
+	}
+
+	@Override
+	protected void changeVisualState(VisualState state) {
```
Comment (leventegal-she) [inline sormas-app/app/src/main/java/de/symeda/sormas/app/component/controls/ControlCheckBoxGroupFieldForMap.java:202]: please remove the TODO , it's just not needed

**Review by leventegal-she — CHANGES_REQUESTED** [review]

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12412

File: sormas-app/app/src/main/java/de/symeda/sormas/app/backend/environment/Environment.java (hunk 2 above)
Comment (sergiupacurariu) [reply sormas-app/app/src/main/java/de/symeda/sormas/app/backend/environment/Environment.java:200]: created

File: sormas-app/app/src/main/java/de/symeda/sormas/app/component/controls/ControlCheckBoxGroupFieldForMap.java (hunk 4 above)
Comment (sergiupacurariu) [reply sormas-app/app/src/main/java/de/symeda/sormas/app/component/controls/ControlCheckBoxGroupFieldForMap.java:162]: moved the code to setEnabled method

File: sormas-app/app/src/main/java/de/symeda/sormas/app/component/controls/ControlCheckBoxGroupFieldForMap.java (hunk 5 above)
Comment (sergiupacurariu) [reply sormas-app/app/src/main/java/de/symeda/sormas/app/component/controls/ControlCheckBoxGroupFieldForMap.java:202]: removed

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=12412

**Review by leventegal-she — APPROVED** [review]

**Pull request merged**